urla958522fc1_spirosikmd_inmdsy_08_tgz-pJ8-TesterAppJ8/src/nl/rug/peerbox/middleware/Sender.java

```diff
@@ -0,0 +1,102 @@
+package nl.rug.peerbox.middleware;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.lang3.concurrent.TimedSemaphore;
+import org.apache.log4j.Logger;
+
+final class Sender {
+
+	private final Queue<Announcement> sentMessagesList = new ConcurrentLinkedQueue<Announcement>();
+	private final BlockingQueue<Announcement> waitingForSendQueue = new ArrayBlockingQueue<Announcement>(
+			1024);
+	private final TimedSemaphore semaphore = new TimedSemaphore(50,
+			TimeUnit.MILLISECONDS, 1);
+
+	private static final Logger logger = Logger.getLogger(Sender.class);
+
+	private final ReliableMulticast group;
+	private Thread thread;
+	private volatile boolean alive = true;
+
+	Sender(ReliableMulticast group) {
+		this.group = group;
+	}
+
+	void start() {
+		thread = new Thread(new Runnable() {
+
+			@Override
+			public void run() {
+				while (alive) {
+					try {
+						// limit bandwidth
+						semaphore.acquire();
+						// get message from queue, if no message in queue wait
+						Announcement message = waitingForSendQueue.take();
+
+						ReliableMulticast.logger.debug("Send: " + message);
+
+						byte[] data = message.toByte();
+						DatagramPacket outgoingPacket = new DatagramPacket(
+								data, data.length, group.getAddress(),
+								group.getPort());
+
+						group.getSocket().send(outgoingPacket);
+						if (message.getCommand() == Announcement.MESSAGE) {
+							if (!sentMessagesList.contains(message)) {
+								sentMessagesList.add(message);
+							}
+						}
+					} catch (IOException e) {
+						alive = false;
+						logger.error(e);
+					} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)